tests/sys_irq_handler/main.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (C) 2019 Gunar Schorcht
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup tests
+ * @{
+ *
+ * @file
+ * @brief       Interrupt handler test application
+ *
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ *
+ * This application demonstrates how to use the interrupt handler thread
+ * module `irq_handler`. For that purpose the main thread simulates interrupts
+ * from 2 interrupt sources with different priorities with a period
+ * of a half second. One interrupt source triggeres a second interrupt before
+ * the interrupts can be handled by the interrupt handler.
+ *
+ * To be able to simulate the interrupts by the main thread, the interrupt
+ * handler thread has to have a lower priority as the main thread for this
+ * example. Otherwise, the interrupts would be handled immediatly when they
+ * are generated by the main thread. Therefore the interrupt handler thread
+ * priority is set to THREAD_PRIORITY_MAIN+1 for this example.
+ * @}
+ */
+
+#include <stdio.h>
+
+#include "irq_handler.h"
+#include "thread.h"
+#include "xtimer.h"
+
+static char some_stack[THREAD_STACKSIZE_MAIN];
+static volatile int i = 0;
```

Comment:
@maribu I know, I remember on the discussion regarding the `volatile` variables also in the scheduler. I don't remember exactly, but I think keyword `volatile` in this case was necessary to avoid the compiler to optimize it out. I will take a look again.

Reply:
Atomic accesses to C11 atomic variables will also not be optimized out, but in addition have a defined memory semantic. E.g. with default memory order (`memory_order_seq_cst`): If `atomic_uint_least32_t x` has the value `0` and `0xffffffff` is added to it, any other concurrently running code (including interrupt service routines) will first read it as `0` and at some point will read it as `0xffffffff`. So all concurrent contexts will observe the same order of changes (it prevents reordering by both compiler *and* CPU) and it will make sure that changes appear atomically. With`volatile` you do not have any ordering guarantees at all, as the compiler will not add CPU instructions where necessary to prevent it from reordering accesses.

Reply:
Let me give you an example how `i++` with `volatile` can fail: With `volatile` every access to `i` is forced to be done on RAM, which on the AVR means two memory accesses (`int` on AVR is 16 bit, the AVR can access 8 bit of RAM at a time). Lets say prior to `i++` the value 255 is stored in `i`. With `i++` it would be possible that first the least significant byte is written to memory (so `i` has temporary value of `0`) and afterwards the msb is written (`i` now has the correct value of 256).

If the check of `i > 0` is executed at the very instance `i` was incremented from 255 to 256 and the lsb has been written, but the msb not, this check will fail.